control-plane/csi-driver/src/bin/node/nodeplugin_svc.rs

```diff
@@ -23,104 +26,100 @@ pub(crate) enum ServiceError {
-                    volume_id: volume_id.to_string(),
-                    error: errmsg,
+                Err(ServiceError::OperationInProgress {
+                    id: uuid.to_string(),
                 })
-            };
-        } else {
-            // mount::find_mount does not return any matches,
-            // for mounts which are bind mounts to block devices
-            // (raw block volume).
-            // It would be incorrect to return the VolumeNotFound error,
-            // if the volume is mounted as a raw block volume on this node.
-            // Use findmnt to work out if volume is mounted as a raw
-            // block, i.e. we get some matches, and return the
-            // BlockDeviceMount error.
-            let mountpaths = findmnt::get_mountpaths(&device_path).context(InternalFailure {
-                volume_id: volume_id.to_string(),
-            })?;
-            if !mountpaths.is_empty() {
-                debug!(
-                    "{} for volume_id :{} : failed for block device",
-                    freeze_op, volume_id
-                );
-                return Err(ServiceError::BlockDeviceMount {
-                    volume_id: volume_id.to_string(),
-                });
             }
-            debug!(
-                "{} for volume_id :{} : failed, cannot find volume",
-                freeze_op, volume_id
-            );
+            false => {
+                inventory.insert(uuid);
+                info!(volume.uuid = %uuid, "New fsfreeze reservation");
+                Ok(Self { uuid })
+            }
         }
     }
-    Err(ServiceError::VolumeNotFound {
-        volume_id: volume_id.to_string(),
-    })
 }
 
-pub(crate) async fn freeze_volume(volume_id: &str) -> Result<(), ServiceError> {
-    fsfreeze(volume_id, "--freeze").await
+impl Drop for VolumeFsfreezeGuard {
+    fn drop(&mut self) {
+        info!(volume.uuid = %self.uuid, "Reservation dropped");
+        let mut hash_set = volume_fsfreeze_serializer().lock().unwrap();
+        hash_set.remove(&self.uuid);
+    }
 }
 
-pub(crate) async fn unfreeze_volume(volume_id: &str) -> Result<(), ServiceError> {
-    fsfreeze(volume_id, "--unfreeze").await
+/// Returns or inits a hash set which stores the uuid's of all volumes which
+/// have fsfreeze operations in-flight.
+/// This can then be used to ensure there is only 1 fsfreeze operation in-flight for a given volume.
+fn volume_fsfreeze_serializer() -> &'static Mutex<HashSet<uuid::Uuid>> {
+    static FSFREEZE_LIMITER: OnceCell<Mutex<HashSet<uuid::Uuid>>> = OnceCell::new();
+
+    FSFREEZE_LIMITER.get_or_init(|| Mutex::new(HashSet::new()))
 }
 
+/// Type of mount.
+pub(crate) enum TypeOfMount {
+    FileSystem,
+    RawBlock,
+}
+
+#[derive(EnumString, Clone, Debug, Eq, PartialEq, Display)]
+#[strum(serialize_all = "lowercase")]
+pub(crate) enum FsFreezeOpt {
+    Freeze,
+    Unfreeze,
+}
+
+// Nix macro to issue FIFREEZE ioctl.
+ioctl_readwrite!(fs_freeze, 'X', 119, i32);
```

Comment:
freeze/unfreeze "low-level" code should be in separate module

Reply:
Moved.